### ISSUE TITLE
Can vendor/src/ become just vendor/?

### ISSUE THREAD

**Issue by peterbourgon** [constabulary/gb#50]
This may have been asked and answered before; if so, forgive the noise.

Given gb is a greenfield implementation of the build tooling, it's not bound to conventions, like that all Go code must live in a directory with an ancestor called src. When looking at gb project structures, vendor/src/ feels stuttery, incongruent with src/, somehow unnecessary. Can vendored code live directly in vendor/, e.g. vendor/github.com/foo/bar/?

**Comment by ChrisHines** [comment]
It also could be a source of confusion. If `gb build` is executed from within a vendor/src/lib directory won't gb pick vendor as the project root instead of vendor/..?

**Comment by Supermighty** [comment]
I like having vendor/src/ because I want to repurpose vendor/ to include all of my 3rd party resources like css and javascript, not just go code.

**Comment by peterbourgon** [comment]
@Supermighty hmm, I think that's a complicating use-case that doesn't need to be supported. You could just as easily put your third-party resources in another directory.

**Comment by Supermighty** [comment]
If `gb` doesn't do anything special with vendor/ then why not put other stuff in it? I would rather have all my external dependencies in one place than litter my project directory with vendor-go/ vendor-css/ vendor-js/ vendor-lua/ or whatever other external libs I am using.

`gb` doesn't need to support anything, it just needs to make a promise to not clobber vendor/ directories. It doesn't need to make the same promise for vendor/src though.

**Comment by ngrilly** [comment]
I agree with @peterbourgon that the shorter the better. **Maybe this is a bad idea**, but as we are discussing this, why not put the dependencies directly in `<project_root>/src`? Packages are naturally namespaced. If the application code lives in `<project_root>/src/myapp`, then it's already separated from the dependencies living in `<project_root>/src/{github.com, bitbucket.org, ...}`.

**Comment by davecheney** [comment]
> I like having vendor/src/ because I want to repurpose vendor/ to include all of my 3rd party resources like css and javascript, not just go code.

@Supermighty in #25 I suggested that `vendor/` would not be a good place for static assets, modulo feelings of minimalism, you can create any directory structure you like in a `gb` project; but I'd strongly encourage you to think of `vendor/` as being _their stuff_ not _your stuff_. Please reopen #25 and explain what 3rd party assets you think are not _your stuff_ and why they must go in `vendor/`

**Comment by davecheney** [comment]
> I agree with @peterbourgon that the shorter the better. But as we are discussing this, why not put the dependencies directly in <project_root>/src? Maybe this is a bad idea, but packages in Go are already naturally namespaced. If the application code lives in <project_root>/src/myapp, then it's already separated from the dependencies living in <project_root>/src/{github.com, bitbucket.org, ...}.

@ngrilly I deliberately chose to separate the projects' code, _your stuff_, from the code that it depends on, _not your stuff_, into it's own directory `vendor/src/`. This was for two reasons

1. political, one of the complaints raised about GOPATH is there is no separation between _my stuff_ and _your stuff_ inside GOPATH. An argument can be made that the url namespacing enforced by `go get` gives users a clear way to identify things that are not _their stuff_; just look a the url, if it's not your github username or org, it's not your stuff -- but some are unconvinced by this argument. 

In short, I'm giving people what they wanted.

2. At a technical level, considering the proposal in #49 it is critical that for any `gb` project, a clear separation of code that represents the project being vendored, and code that that code that is being vendored depends on (what a mouthful).

At a practical level, if you want to create a `gb` project that places all the source of the project, even the dependencies into `$PROJECT/src/` the tool won't stop you doing that. but it's unlikely that tools like `gb-vendor` if rewritten (see #29 ) will support this alternative method of vendoring.

**Comment by davecheney** [comment]
@peterbourgon thanks for raising this. I'm giving this strong consideration as there is a time pressure if the vendor directory does change location.

I want to give some background, and fess up to some features that are not yet implemented. 

At the moment `src/` and `vendor/src/` are hard coded, but it was not my intention that this would be permanent. I had in mind that additional source directories of a project could be added, and auto detected in the same way we auto detect the project root; any directory that matches `$PROJECT/*/src/` is a candidate for searching when resolving import paths to packages. So there would have to be a marker in the path, `src/` to help identify directories when searching.

@ChrisHines you are correct that modeling `vendor/src/` as effectively a gb project inside a gb project would cause confusing results when executing `gb build` inside `$PROJECT/vendor/src/something`; the project root would be reset to `$PROJECT/vendor`. So code would be built, but in the wrong place. This idea of a gb project within a gb project is probably also the result of an incomplete story for how libraries could be packaged as gb projects, see #49 

However, the opposite is also confusing, if we expunged the `src/` directory, then gb build at any point in the project would most likely be able to figure out the correct project root -- but it would _still_ not build anything inside `vendor/src/`, these paths are ignored as part of initial package resolution, they are only referenced if there are unmet transitive dependencies inside the source in `$PROJECT/src/`

**Comment by davecheney** [comment]
To continue the previous thought, on reflection, allowing the flexibility to have multiple vendor directories inside a single project sounded like a good idea, but probably was short sighted,

The problem with auto detecting vendor directories is their order will not be stable, and this directly affects package resolution. As people figure out they don't need to namespace their packages with unique github urls, it is easy to see two packges called "logger", living in different vendor directories usurping each other depending on the order in which they show up in the directory. This instability also would completely screw incremental compilation.

Of course, we could sort the list of vendor directories during discovery, but this feels like a patch on top of a misfeature, and there are number of ways I can see it going wrong with respect to incremental compilation.

So, with that realisation it is clear that the hard coding of the location of the vendor directory will most likely remain hard coded, and that there will not be, for the foreseeable future, support for multiple vendor directories. I think this is an important simplification. Thanks for helping me understand this @peterbourgon.

With all this said, it would appear to be a shoe in that `vendor/src/` can be also simplified to `vendor/` as we don't need the auto detection logic. And In writing my previous response I was coming around to this idea. However there is one big problem; `$GOPATH` compatibility.

@ngrilly  and many others have voiced concerns that moving away from $GOPATH structured source means breaking compatibility with ides, and tools, see #23 and others. My response to those concerns, at worst, GOPATH can be simulated for a gb project by `GOPATH=$PROJECT:$PROJECT/vendor`. If we were to drop the trailing `src/` in the vendor directory we would loose the ability to have a `gb` project impersonate a GOPATH.

So for that reason, at this point, I think we have to keep the `src/` suffix in the vendor directory. And given how long a transition period would be, should gb prove successful in the market, it's likely that `vendor/src/` will remain as a curio permanently.

Anyway, this is what I'm thinking as of today. What have I missed? What have I misunderstood or an not seeing clearly ?

**Comment by ChrisHines** [comment]
I see. Thanks for the clarification. I had not thought carefully enough about what `gb build` from within the vendor tree *should* mean. It does seem rather useless as you point out.

Perhaps the project root auto-detection needs a special case to ignore a `vendor/src` pair when walking up the tree from the working directory to avoid mistaking the vendor tree as a project. In addition, it would be nice if gb finds the project even if the current directory is `<project>/pkg` or `<project>/bin`. I don't know if the code does that at the moment, but the README is not clear on this question. It states:

>A gb project is defined as any directory that contains a src/ subdirectory. gb automatically detects the root of the project by looking at the current working directory and walking backwards until it finds a directory called src/.

Which implies it will only work if you are in a sub-directory of `<project>/src`.

**Comment by peterbourgon** [comment]
> @ngrilly and many others have voiced concerns that moving away from $GOPATH structured source means breaking compatibility with ides, and tools . . .

To be a sort of devil's advocate, I'll submit: the $GOPATH structure, specifically that all buildable code lives under a $GOPATH/src, is a convention baked into the go tool. If gb is starting from fresh foundations, I don't see that it needs to bless this convention for the sake backwards compatibility. That is, it may easily be gb's convention that *your* code lives in $PROJECTROOT/src, and your dependencies live in $PROJECTROOT/vendor.

**Comment by davecheney** [comment]
> To be a sort of devil's advocate, I'll submit: the $GOPATH structure,
specifically that all buildable code lives under a $GOPATH/src, is a
convention baked into the go tool. If gb is starting from fresh
foundations, I don't see that it needs to bless this convention for the
sake backwards compatibility. That is, it may easily be gb's convention
that *your* code lives in $PROJECTROOT/src, and your dependencies live in
$PROJECTROOT/vendor.

I agree with that argument, but am wary of forcing gb users to commit to
dropping any coincidental compatibility with GOPATH based tools. It does
seem like a large cost to pay for three characters.

**Comment by peterbourgon** [comment]
I'm less concerned with the character count, more concerned with the inconsistency in tree depth. What about keeping _their stuff_ in vendor/src/, but moving _my stuff_ to e.g. project/src/?

**Comment by davecheney** [comment]
> I'm less concerned with the character count, more concerned with the
inconsistency in tree depth. What about keeping *their stuff* in
vendor/src/, but moving *my stuff* to e.g. project/src/?

I don't see it as inconsistent. The source for the standard library is in
$GOROOT/src, the source for your project is in $PROJECT/src, and the source
for your project's dependencies is in vendor/src.

**Comment by peterbourgon** [comment]
Well, vendor/src/ is actually $PROJECT/vendor/src/. But, points made, I think.

**Comment by davecheney** [comment]
@chrishines you make a good point that I need to work harder to clarify the for project detection, and have a better sorry for what happens when building inside vendor/ .

Let me take this question on notice and I'll follow up with a new issue when I get back to my desk.

**Comment by davecheney** [comment]
I'm going to mark this closed for the moment.

**Comment by jooon** [comment]
I guess this should be reconsidered again given the new -vendor flag proposal for Go 1.5:

https://groups.google.com/d/msg/golang-dev/74zjMON9glU/4lWCRDCRZg0J

Even if this means breaking $GOPATH compatibility, this will be something new to be compatible with.

**Comment by PSG-Luna** [comment]
I think this'd be a good idea because it'd also remove the ambiguity between vendor directories and project directories, allowing gb to not confuse the two.

**Comment by davecheney** [comment]
That linked thread is announcing support for vendoring dependencies in a
manner compatible with go get. GB doesn't use go get, so there is immediate
change required.
